pysmt/test/omt_examples.py

```diff
@@ -0,0 +1,551 @@
+            elif optimization_type in (OptimizationTypes.LEXICOGRAPHIC, OptimizationTypes.BOXED):
+                goals_number = len(current_goals)
+                assert goals_number > 0, f"{optimization_type.name} needs at least one goal"
+                assert len(goals_oracle_values) == goals_number, f"In {optimization_type.name} optimization the number of goals values must be the same of the number of the given goals: {goals_oracle_values}, {goals_number}"
+            elif optimization_type == OptimizationTypes.PARETO:
+                goals_number = len(current_goals)
+                assert goals_number > 0, f"{optimization_type.name} needs at least one goal"
+                for pareto_goals in goals_oracle_values:
+                    assert isinstance(pareto_goals, (list, tuple)), f"In {optimization_type.name} optimization the goals oracle value must be a list or a tuple of FNode"
+                    assert len(pareto_goals) == goals_number, f"In {optimization_type.name} optimization the goals number of goal values must equal the number of given goals"
+            else:
+                raise NotImplementedError(f"{optimization_type.name} optimization is not supported yet")
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def assertions(self):
+        return self._assertions
+
+    @property
+    def logic(self):
+        return self._logic
+
+    @property
+    def solvable(self):
+        return self._solvable
+
+    @property
+    def goals(self):
+        return self._goals
+
+
+def _fast_examples():
+    yield qf_lia_two_variables_multi_obj_example()
+    yield qf_lia_2_int_2_bools_multiple_objective()
+    yield qf_lra_box_volume()
+    yield qf_lia_box_volume()
+
+
+def _slow_examples():
+    yield test_maximize_revenue()
+
+
+def get_full_example_omt_formuale(environment=None, fast=True, slow=True):
+    """Return a list of OMTTestCases using the given environment."""
+
+    if environment is None:
+        environment = get_env()
+    assert fast or slow, "If neither fast or slow are True this method returns no test cases"
+
+    with environment:
+        omt_examples = []
+        if fast:
+            omt_examples.extend(_fast_examples())
+        if slow:
+            omt_examples.extend(_slow_examples())
+    return omt_examples
+
+
+# method to solve the given examples
+def generate_examples_with_solvers(optimization_examples):
+    """
+    This method takes a list of OMTTestCases and yields all the possible
+    combinations of an OMTTestCase and the name of a solver that support
+    the logic of said test.
+    """
+    for optimization_example in optimization_examples:
+        for solver_name in get_env().factory.all_optimizers(logic=optimization_example.logic):
+            # is_sua_or_incr is passed to the methods to skip the unbound or infinitesimal cases
+            is_sua_or_incr = "_sua" in solver_name or "_incr" in solver_name
```

Comment:
Why not checking the `can_diverge_for_infinitesimal` method (that should be static)?

Reply:
Yes, good idea!

Just 2 notes:

1.  the `can_diverge_for_unbounded_cases` method is not static
2. here we have just the string `solver_name`, so the Optimizer class has to be instantiated.

The cleanest solution I see is:
- make the method static
- make a factory method that returns the class of an optimizer from the name (should I also do it for solvers, just for completeness?)
- use the static method here

Let me know what you think about this

Reply:
I like the solution, but there is no need to touch the factory: the `all_optimizers` method already returns a dict frm solver name to optimizer class :smile: